File: src/main/java/com/github/darylyeung/zentaoplugin/extension/zentao/ZentaoRepository.java
package com.github.darylyeung.zentaoplugin.extension.zentao;

import com.github.darylyeung.zentaoplugin.common.ZentaoConstant;
import com.github.darylyeung.zentaoplugin.extension.zentao.model.*;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.intellij.tasks.Task;
import com.intellij.tasks.TaskRepositoryType;
import com.intellij.tasks.impl.BaseRepository;
import com.intellij.tasks.impl.httpclient.NewBaseRepositoryImpl;
import com.intellij.tasks.impl.httpclient.TaskResponseUtil;
import com.intellij.util.containers.ContainerUtil;
import com.intellij.util.xmlb.annotations.Tag;
import org.apache.commons.collections.CollectionUtils;
import org.apache.http.HttpRequestInterceptor;
import org.apache.http.client.ResponseHandler;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.entity.ContentType;
import org.apache.http.entity.StringEntity;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.net.URISyntaxException;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import static com.github.darylyeung.zentaoplugin.util.ZentaoUtil.GSON;

/**
 * @author Yeung
 * @version v1.0
 * @date 2024-01-10 23:02:18
 */
@Tag("Zentao")
public class ZentaoRepository extends NewBaseRepositoryImpl {

    private String token;
    private ZentaoProduct myCurrentProduct;

    private List<ZentaoProduct> myProducts = null;
    //    private static final Gson GSON = TaskGsonUtil.createDefaultBuilder().create();

    private static final TypeToken<List<ZentaoProduct>> LIST_OF_PRODUCTS_TYPE = new TypeToken<>() {};

    private static final TypeToken<List<ZentaoBug>> LIST_OF_BUGS_TYPE = new TypeToken<>() {};

    public static final ZentaoProduct UNSPECIFIED_PRODUCT = createUnspecifiedProduct();

    private static ZentaoProduct createUnspecifiedProduct() {
        final ZentaoProduct unspecified = new ZentaoProduct() {
            @Override
            public String getName() {
                return "-- all issues created by you --";
            }
        };
        unspecified.setId(-1);
        return unspecified;
    }

    public ZentaoRepository() {
        super();
    }

    public ZentaoRepository(ZentaoRepository other) {
        super(other);
        myCurrentProduct = other.myCurrentProduct;
    }


    @Override
    public @Nullable Task findTask(@NotNull String s) throws Exception {
        return null;
    }

    public ZentaoRepository(TaskRepositoryType type) {
        super(type);
    }

    @Override
    public @NotNull BaseRepository clone() {
        return new ZentaoRepository(this);
    }

    @Override
    public Task[] getIssues(@Nullable String query, int offset, int limit, boolean withClosed) throws Exception {
        final List<ZentaoBug> bugs = fetchBugs((offset / limit) + 1, limit, !withClosed);
        return ContainerUtil.map2Array(bugs, ZentaoTask.class, issue -> new ZentaoTask(this, issue));
    }

    private List<ZentaoBug> fetchBugs(int pageNumber, int pageSize, boolean openedOnly) throws Exception {
        ensureProjectsDiscovered();
        final URIBuilder uriBuilder = new URIBuilder(getBugsUrl())
                .addParameter("page", String.valueOf(pageNumber))
                .addParameter("per_page", String.valueOf(pageSize))
                // Ordering was added in v7.8
                .addParameter("order_by", "updated_at");
        if (openedOnly) {
            // Filtering by state was added in v7.3
            uriBuilder.addParameter("state", "opened");
        }
        final ResponseHandler<ZentaoBugPage> handler = new TaskResponseUtil.GsonSingleObjectDeserializer<>(new Gson(), ZentaoBugPage.class);
        ZentaoBugPage page = getHttpClient().execute(new HttpGet(uriBuilder.build()), handler);
        if (CollectionUtils.isEmpty(page.getBugs())) {
            return Collections.emptyList();
        }
        //  filter my bugs
        return page.getBugs().stream().filter(i -> Objects.equals(i.getAssignedTo().getAccount(), myUsername)).collect(Collectors.toList());
    }

    private String getBugsUrl() {
        if (myCurrentProduct != null && myCurrentProduct != UNSPECIFIED_PRODUCT) {
            return getRestApiUrl("projects", myCurrentProduct.getId(), "issues");
        }
        if (myCurrentProduct == null) {
            int id = myProducts.get(0).getId();
            return getRestApiUrl("products", id, "bugs");
        }
        return getRestApiUrl("bugs");
    }

    @NotNull
    public List<ZentaoProduct> getProducts() {
        try {
            ensureProjectsDiscovered();
        } catch (Exception ignored) {
            return Collections.emptyList();
        }
        return Collections.unmodifiableList(myProducts);
    }

    private void ensureProjectsDiscovered() throws Exception {
        if (myProducts == null) {
            fetchProducts();
        }
    }

    @Override
    public @NotNull String getRestApiPathPrefix() {
        return "/api.php/v1";
    }

    @Nullable
    @Override
    protected HttpRequestInterceptor createRequestInterceptor() {
        if (getToken() == null) {
            return null;
        }
        return (request, context) -> request.addHeader(ZentaoConstant.TOKEN.getCode(), token);
    }

    /**
     * Always forcibly attempt to fetch new projects from server.
     */
    @NotNull
    public List<ZentaoProduct> fetchProducts() throws Exception {
        //  设置Token
        generateToken();
        //  获取产品
        final ResponseHandler<ZentaoProductPage> handler = new TaskResponseUtil.GsonSingleObjectDeserializer<>(GSON, ZentaoProductPage.class);
        final String productUrl = getRestApiUrl("products");
        final URIBuilder productsUrl = new URIBuilder(productUrl);
        final ZentaoProductPage page = getHttpClient().execute(new HttpGet(productsUrl.build()), handler);
        if (Objects.isNull(page) || page.getTotal() < 1) {
            return Collections.emptyList();
        }
        myProducts = page.getProducts();
        return Collections.unmodifiableList(myProducts);
    }

    private void generateToken() throws URISyntaxException, IOException {
        final ResponseHandler<ZentaoToken> handler = new TaskResponseUtil.GsonSingleObjectDeserializer<>(GSON, ZentaoToken.class);
        String tokenUrl = getRestApiUrl("tokens");
        final URIBuilder uriBuilder = new URIBuilder(tokenUrl);
        HttpPost request = new HttpPost(uriBuilder.build());
        request.setEntity(new StringEntity(GSON.toJson(new ZentaoLogin(myUsername, myPassword)), ContentType.APPLICATION_JSON));
        final ZentaoToken token = getHttpClient().execute(request, handler);
        if (token != null) {
            setToken(token.getToken());
        }
    }

    public ZentaoProduct getCurrentProduct() {
        return myCurrentProduct;
    }

    public String getToken() {
        return token;
    }

    public void setToken(String token) {
        this.token = token;
    }
}

File: src/main/kotlin/com/github/darylyeung/zentaoplugin/extension/zentao/ZentaoRepository.kt
package com.github.darylyeung.zentaoplugin.extension.zentao

import com.github.darylyeung.zentaoplugin.common.Constant
import com.github.darylyeung.zentaoplugin.extension.zentao.model.*
import com.intellij.openapi.diagnostic.thisLogger
import com.intellij.tasks.Task
import com.intellij.tasks.impl.BaseRepository
import com.intellij.tasks.impl.httpclient.NewBaseRepositoryImpl
import com.intellij.util.containers.ContainerUtil
import com.intellij.util.xmlb.annotations.Tag
import io.ktor.client.plugins.observer.*
import io.ktor.client.request.*
import io.ktor.http.*
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.apache.http.HttpRequestInterceptor
import org.jetbrains.annotations.NotNull
import org.jetbrains.annotations.Nullable

/**
 * @author Yeung
 * @version v1.0
 * @date 2024-01-18 22:32:15
 */
@Tag("Zentao")
class ZentaoRepository : NewBaseRepositoryImpl {
    private var token: String? = null
    private var myCurrentProduct: ZentaoProduct? = null
    private var myProducts: List<ZentaoProduct>? = null

    constructor() : super()

    constructor(other: ZentaoRepository) : super(other) {
        myCurrentProduct = other.myCurrentProduct
    }

    constructor(type: ZentaoRepositoryType) : super(type)

    override fun findTask(s: String): Task? {
        return null
    }

    override fun clone(): BaseRepository {
        return ZentaoRepository(this)
    }

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    @Throws(Exception::class)
    override fun getIssues(query: String?, offset: Int, limit: Int, withClosed: Boolean): Array<out ZentaoTask> {
        val bugs = fetchBugs((offset / limit) + 1, limit, !withClosed)
        return ContainerUtil.map2Array(bugs, ZentaoTask::class.java) { issue -> ZentaoTask(this, issue) }
    }

    @Throws(Exception::class)
    private fun fetchBugs(pageNumber: Int, pageSize: Int, openedOnly: Boolean): List<ZentaoBug> {
        ensureProjectsDiscovered()
        if (myCurrentProduct != null) {
            return getBugs(getRestApiUrl("products", myCurrentProduct?.id, "bugs"))
        } else {
            val bugs = mutableListOf<ZentaoBug>()
            myProducts?.forEach { product ->
                bugs.plus(getBugs(getRestApiUrl("products", product.id, "bugs")))
            }
            return bugs
        }
    }


    private fun getBugs(url: String): List<ZentaoBug> {
        val response = OkHttpClient().newCall(
            Request.Builder().url(url)
                .header(Constant.TOKEN_KEY.value, token.toString()).build()
        ).execute()
        if (!response.isSuccessful) {
            return emptyList()
        }
        val result = response.body?.string().toString()
        thisLogger().info("fetch bug Successful. Response: $result")
        val page = json.decodeFromString<ZentaoBugPage>(result)
        return if (page.bugs.isEmpty()) {
            emptyList()
        } else {
            page.bugs.filter { i -> i.assignedTo.account == myUsername }.toList()
        }
    }

    @NotNull
    fun getProducts(): List<ZentaoProduct> {
        return try {
            ensureProjectsDiscovered()
            myProducts.orEmpty()
        } catch (ignored: Exception) {
            emptyList()
        }
    }

    @Throws(Exception::class)
    private fun ensureProjectsDiscovered() {
        if (myProducts == null) {
            myProducts = fetchProducts()
        }
    }

    override fun getRestApiPathPrefix(): String {
        return "/api.php/v1"
    }

    @Nullable
    override fun createRequestInterceptor(): HttpRequestInterceptor? {
        return if (token == null) {
            null
        } else HttpRequestInterceptor { request, _ -> request.addHeader(Constant.TOKEN_KEY.value, token) }
    }

    @NotNull
    @Throws(Exception::class)
    fun fetchProducts(): List<ZentaoProduct> {
        generateToken()

        val response = OkHttpClient().newCall(
            Request.Builder()
                .url(getRestApiUrl("products"))
                .header(Constant.TOKEN_KEY.value, token.toString()).build()
        ).execute()
        if (!response.isSuccessful) {
            return emptyList()
        }
        val result = response.body?.string().toString()
        thisLogger().info("fetch product Successful. Response: $result")
        val page = json.decodeFromString<ZentaoProductPage>(result)
        return if (page.products.isEmpty()) {
            emptyList()
        } else {
            page.products
        }
    }

    @Throws(Exception::class)
    fun generateToken() {
        val paramJson = "{\"account\":\"$myUsername\",\"password\":\"$myPassword\"}"
        val body = paramJson.toRequestBody("application/json; charset=utf-8".toMediaTypeOrNull())
        val request = Request.Builder().url(getRestApiUrl("tokens"))
            .header("Content-Type", "application/json; charset=utf-8")
            .post(body).build()
        val response = OkHttpClient().newCall(request).execute()
        if (!response.isSuccessful) {
            return
        }
        val result = response.body?.string().toString()
        thisLogger().info("fetch token Successful. Response: $result")
        if (result.isEmpty()) {
            return
        }
        val token = json.decodeFromString<ZentaoToken>(result)
        setToken(token.token)
    }

    fun getCurrentProduct(): ZentaoProduct? {
        return myCurrentProduct
    }

    fun getToken(): String? {
        return token
    }

    fun setToken(token: String?) {
        this.token = token
    }
}